### PR TITLE
Re-implement DockerOnceRetentionStrategy

### PR DESCRIPTION
**Pull request by pjdarton**
We have existing bug reports that suggest that DockerOnceRetentionStrategy is killing containers while they're busy, as builds are failing at the same time a container closes.
i.e. even though the retention strategy is being told "it's over", it doesn't seem to be over.

This PR is an attempt to re-implement the retention strategy to keep track of activity and only close the container down once we're idle.
...however, it's marked as a "work in progress" as it's going to require some non-trivial diagnostics to confirm if this is a good idea or not...

TL;DR: If you're experiencing problems with docker nodes terminating while in use, please do see if this fixes anything.

### CODE REVIEW

**Comment by nalle** [comment]
Hi, we've run this fix for quite some time and it works great... however, someone updated our instances to 1.2.6 and now we're having the issues again.

Is it possible to rebuild with the new version?

**Comment by pjdarton** [comment]
You'll find the CI build of this PR here: https://ci.jenkins.io/job/Plugins/job/docker-plugin/job/PR-859/

There's a build of "unreleased version of the plugin + this PR" here: https://ci.jenkins.io/job/Plugins/job/docker-plugin/job/PR-859/11/artifact/io/jenkins/docker/docker-plugin/1.2.7-rc1003.669690129505/

**Comment by nalle** [comment]
Thanks! Much appreciated :-)
